### PR TITLE
Extract title from filename successfully when no date.

### PR DESCRIPTION
**Pull request by jekyllbot**
PR automatically created for @envygeeks .
Closes #4191 with a slight cleanup `Document#post_read`.

### CODE REVIEW

File: lib/jekyll/document.rb [hunk 1]
```diff
@@ -306,21 +306,21 @@ def read(opts = {})
...
-      populate_tags
+      if relative_path =~ DATE_FILENAME_MATCHER
+        cats, date, slug, ext = $1, $2, $3, $4
+        merge_data!("date" => date) if !data['date'] || data['date'].to_i == site.time.to_i
+        merge_data!("slug" => slug, "ext"  => ext)
 
-      if generate_excerpt?
-        data['excerpt'] ||= Jekyll::Excerpt.new(self)
+      elsif DATELESS_FILENAME_MATCHER =~ relative_path
+        cats, slug, ext = $1, $2, $3
```
Comment (parkr) [inline lib/jekyll/document.rb]: Where do you merge this data? Did you mean to put the slug in `merge_data!`?

Comment (envygeeks) [inline lib/jekyll/document.rb]: @parkr I don't, nice catch, I meant to actually move that down like I did with the `titleize`.

Comment (parkr) [inline lib/jekyll/document.rb]: I wonder how your test still passed then?

Comment (envygeeks) [inline lib/jekyll/document.rb]: @parkr that's something @rebornix and we have to answer, I'll look into that, it seems like we have gaps in the testing that need to be addressed.

Comment (rebornix) [inline lib/jekyll/document.rb]: I deleted the merge data line by mistake when attempting to resolve another test issue, seems it has nothing to do with it but I forgot to bring it back.

`slug: Utils.slugify(data['slug']) || Utils.slugify(basename_without_ext),` for url_placeholders in document.rb is stopping us from test failure. For dateless filename, I suppose basename_without_ext equals slug.

**Comment by parkr** [comment]
Needs a rebase and to have the above problem addressed.

**Comment by envygeeks** [comment]
Attaching it to 3.1 so I get  a reminder to work on it before the release date.

**Comment by envygeeks** [comment]
Hulk: +backlog +jekyll +v3.1

**Comment by parkr** [comment]
This needs a rebase, now.

**Comment by parkr** [comment]
@envygeeks This look OK to you?

**Comment by parkr** [comment]
@jekyllbot: merge +bug

**Pull request merged**